File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        a = []
        sorted_nums = sorted(nums)
        i = 0
        while i < len(sorted_nums):
            j = i + 1
            k = len(sorted_nums)-1
            while j < k and j < len(sorted_nums)-1:
                total = sorted_nums[i] + sorted_nums[j] + sorted_nums[k]
                if(total == 0):
                    a.append([sorted_nums[i], sorted_nums[j], sorted_nums[k]])
                while total == 0 and j + 1 < len(sorted_nums)-1 and  sorted_nums[j] == sorted_nums[j + 1]:
                    j += 1
                if(total > 0):
                    k -= 1
                else:
                    j += 1
            while i + 1 < len(sorted_nums)-1 and  sorted_nums[i] == sorted_nums[i + 1]:
                    i += 1
            i += 1
            
        return a


        

File: 0015-3sum/0015-3sum.ts
function threeSum(nums: number[]): number[][] {
  const res: number[][] = [];
  const n = nums.length;
  if (n < 3) return res;

  nums.sort((a, b) => a - b);

  for (let i = 0; i < n - 2; i++) {
    // skip duplicate anchors
    if (i > 0 && nums[i] === nums[i - 1]) continue;

    // pruning: if the smallest possible sum > 0, we can stop
    const minSum = nums[i] + nums[i + 1] + nums[i + 2];
    if (minSum > 0) break;

    // pruning: if the largest possible sum < 0, move i forward
    const maxSum = nums[i] + nums[n - 2] + nums[n - 1];
    if (maxSum < 0) continue;

    let l = i + 1;
    let r = n - 1;

    while (l < r) {
      const sum = nums[i] + nums[l] + nums[r];

      if (sum === 0) {
        res.push([nums[i], nums[l], nums[r]]);
        // move past duplicates on both sides
        const leftVal = nums[l], rightVal = nums[r];
        while (l < r && nums[l] === leftVal) l++;
        while (l < r && nums[r] === rightVal) r--;
      } else if (sum < 0) {
        l++;
      } else {
        r--;
      }
    }
  }

  return res;
}
